golang/x/relay/keeper/requests.go

```diff
@@ -0,0 +1,139 @@
+package keeper
+
+import (
+	"encoding/binary"
+	"encoding/json"
+
+	btcspv "github.com/summa-tx/bitcoin-spv/golang/btcspv"
+	"github.com/summa-tx/relays/golang/x/relay/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func (k Keeper) emitProofRequest(ctx sdk.Context, pays, spends []byte, paysValue, id uint64) {
+	ctx.EventManager().EmitEvent(types.NewProofRequestEvent(pays, spends, paysValue, id))
+}
+
+func (k Keeper) getRequestStore(ctx sdk.Context) sdk.KVStore {
+	return k.getPrefixStore(ctx, types.RequestStorePrefix)
+}
+
+func (k Keeper) hasRequest(ctx sdk.Context, id uint64) bool {
+	// convert id to bytes
+	idBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(idBytes, id)
+
+	store := k.getRequestStore(ctx)
+	return store.Has(idBytes)
+}
+
+func (k Keeper) setRequest(ctx sdk.Context, spends []byte, pays []byte, paysValue uint64, numConfs uint8) sdk.Error {
+	store := k.getRequestStore(ctx)
+
+	spendsDigest := btcspv.Hash256(spends)
+	paysDigest := btcspv.Hash256(pays)
+
+	request := types.ProofRequest{
+		Spends:      spendsDigest,
+		Pays:        paysDigest,
+		PaysValue:   paysValue,
+		ActiveState: true,
+		NumConfs:    numConfs,
+	}
+
+	// When a new request comes in, get the id and use it to store request
+	id := k.getID(ctx)
+
+	buf, err := json.Marshal(request)
+	if err != nil {
+		return types.ErrMarshalJSON(types.DefaultCodespace)
+	}
+	store.Set(id, buf)
+
+	// Increment the ID
+	k.incrementID(ctx)
+
+	// Emit Proof Request event
+	numID := binary.BigEndian.Uint64(id)
+	k.emitProofRequest(ctx, pays, spends, request.PaysValue, numID)
+	return nil
+}
+
+func (k Keeper) getRequest(ctx sdk.Context, id uint64) (types.ProofRequest, sdk.Error) {
+	store := k.getRequestStore(ctx)
+
+	// convert id to bytes
+	idBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(idBytes, id)
+
+	hasRequest := k.hasRequest(ctx, id)
+	if !hasRequest {
+		return types.ProofRequest{}, types.ErrUnknownRequest(types.DefaultCodespace)
+	}
+	buf := store.Get(idBytes)
+	var request types.ProofRequest
+	json.Unmarshal(buf, &request)
+	return request, nil
+}
+
+func (k Keeper) incrementID(ctx sdk.Context) {
+	store := k.getRequestStore(ctx)
+	// get id
+	id := k.getID(ctx)
+	// convert id to uint64 and add 1
+	newID := binary.BigEndian.Uint64(id) + 1
+	// convert back to bytes and store
+	b := make([]byte, 8)
+	binary.BigEndian.PutUint64(b, newID)
+	store.Set([]byte(types.RequestID), b)
+}
+
+func (k Keeper) getID(ctx sdk.Context) []byte {
```

Comment:
is this the lowest unused ID or the highest used ID?

Might be good to name it `getNextID` or `getLatestID` to make it clear

Reply:
Hm, well I had two functions: `getID` and `incrementID`.  I could probably just combine those into a `getNextID` function.

Reply:
Yeah, I like that better

Reply:
Yeah, I like that better

Reply:
hmmm. the reason I'd avoid doing this is that it makes it difficult to learn the next ID without consuming the next ID. suppose we wanted to add a `QueryGetNextID` in the future

Reply:
What would be the purpose of knowing the next id if you weren't going to use it?

Reply:
Useful for the off-chain software that manages filling requests, so it knows if it had them all or if it needs to retrieve more

Reply:
Hm, so in that case, would it be more useful to know the last used id or the next id? When I stored a request, I incremented the id when I was done.  So getID would technically give the next id to be used.  But I could easily change that. ```func (k Keeper) setRequest(ctx sdk.Context, spends []byte, pays []byte, paysValue uint64, numConfs uint8) sdk.Error {
	store := k.getRequestStore(ctx)

	spendsDigest := btcspv.Hash256(spends)
	paysDigest := btcspv.Hash256(pays)

	request := types.ProofRequest{
		Spends:      spendsDigest,
		Pays:        paysDigest,
		PaysValue:   paysValue,
		ActiveState: true,
		NumConfs:    numConfs,
	}

	// When a new request comes in, get the id and use it to store request
	id := k.getID(ctx)

	buf, err := json.Marshal(request)
	if err != nil {
		return types.ErrMarshalJSON(types.DefaultCodespace)
	}
	store.Set(id, buf)

	// Increment the ID
	k.incrementID(ctx)

	// Emit Proof Request event
	numID := binary.BigEndian.Uint64(id)
	k.emitProofRequest(ctx, pays, spends, request.PaysValue, numID)
	return nil
}```

Reply:
No preference, so long as it's clearly documented and named :(

Reply:
Ok, I separated them again, named it `getNextID` and added documentation.